pkg/sinks/forward/forward.go

```diff
@@ -357,76 +357,120 @@ func (df *DataForward) ackFromBuffer(ctx context.Context, offsets []isb.Offset)
+					needRetry = true
+
+					// TODO(Retry-Sink) : Propagate the retry-count?
+					// we retry only failed message
+					failedMessages = append(failedMessages, msg)
+
+					metrics.WriteMessagesError.With(map[string]string{metrics.LabelVertex: df.vertexName, metrics.LabelPipeline: df.pipelineName, metrics.LabelVertexType: string(dfv1.VertexTypeSink), metrics.LabelVertexReplicaIndex: strconv.Itoa(int(df.vertexReplica)), metrics.LabelPartitionName: sinkWriter.GetName()}).Inc()
+					// a shutdown can break the blocking loop caused due to InternalErr
+					if ok, _ := df.IsShuttingDown(); ok {
+						metrics.PlatformError.With(map[string]string{metrics.LabelVertex: df.vertexName, metrics.LabelPipeline: df.pipelineName, metrics.LabelVertexType: string(dfv1.VertexTypeSink), metrics.LabelVertexReplicaIndex: strconv.Itoa(int(df.vertexReplica))}).Inc()
+						return true, fmt.Errorf("writeToSink failed, Stop called while stuck on an internal error with failed messages:%d, %v", len(failedMessages), errs)
+					}
+				} else {
+					writeCount++
+					writeBytes += float64(len(msg.Payload))
+					// we support write offsets only for jetstream
+					if _writeOffsets != nil {
+						writeOffsets = append(writeOffsets, _writeOffsets[idx])
+					}
 				}
 			}
+			// set messages to failedMessages, in case of success this should be empty
+			// While checking for retry we see the length of the messages left
+			messagesToTry = failedMessages
+			if needRetry {
+				df.opts.logger.Errorw("Retrying failed messages",
+					zap.Any("errors", errorArrayToMap(errs)),
+					zap.String(metrics.LabelPipeline, df.pipelineName),
+					zap.String(metrics.LabelVertex, df.vertexName),
+					zap.String(metrics.LabelPartitionName, sinkWriter.GetName()),
+				)
+				return false, nil
+			}
+			return true, nil
+		})
+		// Forced shutdown
+		// TODO(Retry-Sink): Check for ctx done? That should be covered in shutdown
+		if ok, _ := df.IsShuttingDown(); err != nil && ok {
+			return nil, nil, err
 		}
-
-		if needRetry {
-			df.opts.logger.Errorw("Retrying failed messages",
-				zap.Any("errors", errorArrayToMap(errs)),
-				zap.String(metrics.LabelPipeline, df.pipelineName),
-				zap.String(metrics.LabelVertex, df.vertexName),
-				zap.String(metrics.LabelPartitionName, sinkWriter.GetName()),
-			)
-			// set messages to failed for the retry
-			messages = failedMessages
-			// TODO: implement retry with backoff etc.
-			time.Sleep(df.opts.retryInterval)
-		} else {
-			break
+		// Check if we still have messages left to be processed
+		if len(messagesToTry) > 0 {
+			df.opts.logger.Infof("%d tries exhausted in sink, messagesLeft %d, Next strategy %s",
+				backoffCond.Steps, len(messagesToTry), failStrategy)
+			// Check what is the failure strategy to be followed after retry exhaustion
+			switch failStrategy {
+			case dfv1.OnFailRetry:
+				// If on failure, we keep on retrying then lets continue the loop and try all again
+				continue
+			case dfv1.OnFailFallback:
```

Comment:
add metrics for Fallback?

Reply:
Added seperate metrics for FB on vertex and monovertex